### PR TITLE
makefile: fix build failure on MinGW

### PR DESCRIPTION
**Pull request by starwing**
libuv build fail with MinGW on windows:

```
SW@SW-PC /d/Work/Code/libuv (master)
$ make -f Makefile.mingw
g++    -c -o src/fs-poll.o src/fs-poll.c
src/fs-poll.c: In function 'int uv_fs_poll_start(uv_fs_poll_t*, uv_fs_poll_cb, c
onst char*, unsigned int)':
src/fs-poll.c:69:37: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = calloc(1, sizeof(*ctx) + len);
                                     ^
src/fs-poll.c: In function 'int uv_fs_poll_stop(uv_fs_poll_t*)':
src/fs-poll.c:103:7: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = handle->poll_ctx;
       ^
<builtin>: recipe for target 'src/fs-poll.o' failed
make: *** [src/fs-poll.o] Error 1
```

changed the build line, allow user to add CFLAGS (use EXTRA_CFLAGS in makefile for libuv itself), and add a clean target.

### CODE REVIEW

File: Makefile.mingw [hunk 1]
```diff
@@ -14,14 +14,14 @@
...
 
-CFLAGS += -Wall \
-          -Wextra \
-          -Wno-unused-parameter \
-          -Iinclude \
-          -Isrc \
-          -Isrc/win \
-          -DWIN32_LEAN_AND_MEAN \
-          -D_WIN32_WINNT=0x0600
+EXTRA_CFLAGS = -Wall \
```
Comment (saghul) [inline Makefile.mingw]: Why do you need this instead of CFLAGS += ? We are not redefining CFLAGS, just extending.

**Comment by saghul** [comment]
How do these changes fix your compilation issue? Also, what gcc version does your MinGW have? I was under the impression that those casts should be automagic :-O

**Comment by starwing** [comment]
this is my usecase:

```

SW@SW-PC /d/Work/Sources/libuv (master)
$ diff Makefile.mingw{,.new}
17,24c17,24
< CFLAGS += -Wall \
<           -Wextra \
<           -Wno-unused-parameter \
<           -Iinclude \
<           -Isrc \
<           -Isrc/win \
<           -DWIN32_LEAN_AND_MEAN \
<           -D_WIN32_WINNT=0x0600
---
> EXTRA_CFLAGS = -Wall \
>                -Wextra \
>                -Wno-unused-parameter \
>                -Iinclude \
>                -Isrc \
>                -Isrc/win \
>                -DWIN32_LEAN_AND_MEAN \
>                -D_WIN32_WINNT=0x0600
72a73,75
> clean:
>       -$(RM) $(OBJS) libuv.a
>
76,77c79,80
< # FIXME(bnoordhuis) Don't rebuild everything when a source file changes.
< $(OBJS): $(OBJS:.o=.c) $(INCLUDES)
---
> $(OBJS): %.o : %.c $(INCLUDES)
>       $(CC) $(CFLAGS) $(EXTRA_CFLAGS) -c -o $@ $<

SW@SW-PC /d/Work/Sources/libuv (master)
$ gcc -v
Using built-in specs.
COLLECT_GCC=d:\MinGW\bin\gcc.exe
COLLECT_LTO_WRAPPER=d:/mingw/bin/../libexec/gcc/i686-w64-mingw32/4.8.2/lto-wrapp
er.exe
Target: i686-w64-mingw32
Configured with: ../../sources/gcc-4.8.2/configure --prefix=/d/temp/dest --build
=i686-w64-mingw32 --host=i686-w64-mingw32 --target=i686-w64-mingw32 --with-nativ
e-system-header-dir=/d/temp/dest/i686-w64-mingw32 --enable-languages=c,c++ --dis
able-multilib --disable-lto --enable-graphite --enable-cloog-backend=isl --disab
le-shared --disable-bootstrap --disable-nls --disable-werror --disable-symvers -
-disable-win32-registry --disable-isl-version-check --disable-cloog-version-chec
k --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-checking=release --
enable-sjlj-exceptions --with-gnu-as --with-gnu-ld --with-tune=core-avx-i --with
-system-zlib --with-libiconv-prefix=/d/temp/prereq --with-zlib=/d/temp/prereq --
with-gmp=/d/temp/prereq --with-mpfr=/d/temp/prereq --with-mpc=/d/temp/prereq --w
ith-cloog=/d/temp/prereq --with-pkgversion='Built by StarWing'
Thread model: win32
gcc version 4.8.2 (Built by StarWing)

SW@SW-PC /d/Work/Sources/libuv (master)
$ uname -a
MINGW32_NT-6.1 SW-PC 1.0.12(0.46/3/2) 2012-07-05 14:56 i686 unknown

SW@SW-PC /d/Work/Sources/libuv (master)
$ make -v
GNU Make 4.0
Built for Windows32
Copyright (C) 1988-2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

SW@SW-PC /d/Work/Sources/libuv (master)
$ make -f Makefile.mingw
g++    -c -o src/fs-poll.o src/fs-poll.c
src/fs-poll.c: In function 'int uv_fs_poll_start(uv_fs_poll_t*, uv_fs_poll_cb, c
onst char*, unsigned int)':
src/fs-poll.c:69:37: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = calloc(1, sizeof(*ctx) + len);
                                     ^
src/fs-poll.c: In function 'int uv_fs_poll_stop(uv_fs_poll_t*)':
src/fs-poll.c:103:7: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = handle->poll_ctx;
       ^
<builtin>: recipe for target 'src/fs-poll.o' failed
make: *** [src/fs-poll.o] Error 1

SW@SW-PC /d/Work/Sources/libuv (master)
$ make -f makefile.mingw CFLAGS="what ever"
g++    -c -o src/fs-poll.o src/fs-poll.c
src/fs-poll.c: In function 'int uv_fs_poll_start(uv_fs_poll_t*, uv_fs_poll_cb, c
onst char*, unsigned int)':
src/fs-poll.c:69:37: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = calloc(1, sizeof(*ctx) + len);
                                     ^
src/fs-poll.c: In function 'int uv_fs_poll_stop(uv_fs_poll_t*)':
src/fs-poll.c:103:7: error: invalid conversion from 'void*' to 'poll_ctx*' [-fpe
rmissive]
   ctx = handle->poll_ctx;
       ^
<builtin>: recipe for target 'src/fs-poll.o' failed
make: *** [src/fs-poll.o] Error 1

SW@SW-PC /d/Work/Sources/libuv (master)
$ make -f Makefile.mingw.new CFLAGS="-O2"
gcc -O2 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/fs-poll.o src/fs-poll.c
gcc -O2 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/inet.o src/inet.c
gcc -O2 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/uv-common.o src/uv-common.c
gcc -O2 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/version.o src/version.c
.......
```

**Comment by saghul** [comment]
That's not what I meant :-) What I wanted to know is why you need the EXTRA_CFLAGS, and using += with CFLAGS is not enough for you.

**Comment by starwing** [comment]
I don't know why it's not work, but it's really can not work at make 4.0 on Windows. you can not use += to add flags if you specific it on command line, and CC, CFLAGS just doesn't work when you use implicit rule.

this also fix the issue about compile all source if you changed one, as @bnoordhuis metioned on comment.

**Comment by starwing** [comment]
rebased to current header.

**Comment by saghul** [comment]
OK, I just tested this myself and I;m not seeing the problem you mention with CFLAGS +=. I'm using bash.exe, and I did `export CFLAGS="-Wfoo"` and compilation failed because of the unrecognized option. I'm using gcc 4.8.1 and make 3.81. Is this problem only visible with make 4.0?

**Comment by starwing** [comment]
I'll try you ways (using env-vars, but not command options), and then try make 3.81, Thanks for your informations :)

**Comment by starwing** [comment]
@saghul I have tried your ways on make 3.82, it can work, but which I want is like this:

```
make -f makefile.mingw CFLAGS="-O3"
```

by the way, the make 3.82 can use implicit rule correctly, but not make 4.0, I'm now clone git repo and build a latest 4.0 one and try :(

**Comment by starwing** [comment]
rebased to least commit.

**Comment by saghul** [comment]
The CFLAGS thing seems to be a problem with your particular environment, it works for me using MinGW installed straight from the website and updated to the latest version using mingw-get. I'm using the builtin bash shell.

```
saghul@foxtrot /c/Users/saghul/src/libuv
$ CFLAGS="-Wfoo" make -f Makefile.mingw
gcc -Wfoo -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_
LEAN_AND_MEAN -D_WIN32_WINNT=0x0600   -c -o src/fs-poll.o src/fs-poll.c
gcc.exe: error: unrecognized command line option '-Wfoo'
make: *** [src/fs-poll.o] Error 1

saghul@foxtrot /c/Users/saghul/src/libuv
$ bash --version
GNU bash, version 3.1.17(1)-release (i686-pc-msys)
Copyright (C) 2005 Free Software Foundation, Inc.

saghul@foxtrot /c/Users/saghul/src/libuv
$
```

The rest of the patch looks ok and works, but I cannot merge the CFLAGS part since I don't see any problem with it. Can you please split up the patch?

**Comment by starwing** [comment]
@saghul But I want to use CFLAGS on make argument, like this:

```
make -f makefile.mingw CFLAGS=barbarbar
```

rename CFLAGS to other name can make this use. maybe we should use some short names? what about UV_CFLAGS?

**Comment by saghul** [comment]
Look at my previous comment, this works:

```
$ CFLAGS="-Wfoo" make -f Makefile.mingw
```

**Comment by starwing** [comment]
okay, I will revert this change.

**Comment by starwing** [comment]
@saghul ready to merge, has some LGTM?

**Comment by saghul** [comment]
LGTM. Can you modify the commit message mentioning what was really fixed? Thanks!

**Comment by starwing** [comment]
Okay, I will change that.

**Comment by starwing** [comment]
Is this message enough, or need more details?

**Comment by saghul** [comment]
Looks good @starwing, have you signed the CLA?

**Comment by saghul** [comment]
I see you signed the CLA. Thanks @starwing! It landed in https://github.com/joyent/libuv/commit/29ed404138e1b351cbed1add1781a218fefc3b01

It took us a while, but I did learn a thing or two in the process! :-)

**Pull request closed without merging**